File: server/internal/auth/authdb/webauthn.go
package authdb

import (
	"context"
	"errors"
	"time"

	"tallyo/internal/auth/authtypes"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"

	"github.com/samber/lo"
)

func (s *Store) CreateWebAuthnCredential(ctx context.Context, credential authtypes.WebAuthnCredential) error {
	err := s.q.CreateWebAuthnCredential(ctx, dbgen.CreateWebAuthnCredentialParams{ID: credential.ID, UserID: credential.UserID, Name: credential.Name, Credential: credential.CredentialJSON})
	return wrapDB(err, "create webauthn credential")
}

func (s *Store) ConsumeWebAuthnRegistration(ctx context.Context, credential authtypes.WebAuthnCredential) error {
	txCtx, err := s.BeginTX(ctx)
	if err != nil {
		return err
	}
	if err := s.CreateWebAuthnCredential(txCtx, credential); err != nil {
		return errors.Join(err, s.Rollback(txCtx))
	}
	if err := s.DeleteWebAuthnRegistration(txCtx, credential.UserID); err != nil {
		return errors.Join(err, s.Rollback(txCtx))
	}
	return s.Commit(txCtx)
}

func (s *Store) WebAuthnCredentialsByUserID(ctx context.Context, userID int64) ([]authtypes.WebAuthnCredential, error) {
	rows, err := s.q.WebAuthnCredentialsByUserID(ctx, dbgen.WebAuthnCredentialsByUserIDParams{UserID: userID})
	return dbutil.MapRows(rows, err, webAuthnCredentialFromRow)
}

func (s *Store) UpdateWebAuthnCredential(ctx context.Context, id, credentialJSON string, lastUsedAt time.Time) error {
	lastUsedAtUTC := lastUsedAt.UTC()
	err := s.q.UpdateWebAuthnCredential(ctx, dbgen.UpdateWebAuthnCredentialParams{Credential: credentialJSON, LastUsedAt: &lastUsedAtUTC, ID: id})
	return wrapDB(err, "update webauthn credential")
}

func (s *Store) RenameWebAuthnCredential(ctx context.Context, id string, userID int64, name string) error {
	rows, err := s.q.RenameWebAuthnCredential(ctx, dbgen.RenameWebAuthnCredentialParams{Name: name, ID: id, UserID: userID})
	return wrapAffectedRows(err, rows, "rename webauthn credential")
}

func (s *Store) DeleteWebAuthnCredential(ctx context.Context, id string, userID int64) error {
	rows, err := s.q.DeleteWebAuthnCredential(ctx, dbgen.DeleteWebAuthnCredentialParams{ID: id, UserID: userID})
	return wrapAffectedRows(err, rows, "delete webauthn credential")
}

func (s *Store) SaveWebAuthnRegistration(ctx context.Context, registration authtypes.WebAuthnRegistration) error {
	err := s.q.SaveWebAuthnRegistration(ctx, dbgen.SaveWebAuthnRegistrationParams{UserID: registration.UserID, Name: registration.Name, Session: registration.SessionJSON, ExpiresAt: registration.ExpiresAt.UTC()})
	return wrapDB(err, "save webauthn registration")
}

func (s *Store) WebAuthnRegistration(ctx context.Context, userID int64) (authtypes.WebAuthnRegistration, error) {
	row, err := s.q.WebAuthnRegistration(ctx, dbgen.WebAuthnRegistrationParams{UserID: userID})
	if err != nil {
		return authtypes.WebAuthnRegistration{}, err
	}
	return webAuthnRegistrationFromRow(row), nil
}

func (s *Store) DeleteWebAuthnRegistration(ctx context.Context, userID int64) error {
	rows, err := s.q.DeleteWebAuthnRegistration(ctx, dbgen.DeleteWebAuthnRegistrationParams{UserID: userID})
	return wrapAffectedRows(err, rows, "delete webauthn registration")
}

func webAuthnCredentialFromRow(row dbgen.WebauthnCredential) authtypes.WebAuthnCredential {
	return authtypes.WebAuthnCredential{ID: row.ID, UserID: row.UserID, Name: row.Name, CredentialJSON: row.Credential, CreatedAt: row.CreatedAt, LastUsedAt: lo.FromPtr(row.LastUsedAt)}
}

func webAuthnRegistrationFromRow(row dbgen.WebauthnRegistration) authtypes.WebAuthnRegistration {
	return authtypes.WebAuthnRegistration{UserID: row.UserID, Name: row.Name, SessionJSON: row.Session, ExpiresAt: row.ExpiresAt}
}

File: server-rs/src/auth/store/webauthn.rs
use anyhow::{Context, Result};
use chrono::{DateTime, Utc};

use crate::database::{queries, with_tx};

use super::{Store, WebAuthnCredential, WebAuthnRegistration, affected};

impl Store {
    pub async fn consume_webauthn_registration(&self, credential: &WebAuthnCredential) -> Result<()> {
        let credential = credential.clone();
        with_tx(&self.pool, |transaction| {
            Box::pin(async move {
                queries::create_web_authn_credential(
                    &mut **transaction,
                    queries::CreateWebAuthnCredentialParams {
                        id: &credential.id,
                        user_id: credential.user_id,
                        name: &credential.name,
                        credential: &credential.credential,
                    },
                )
                .await?;
                affected(
                    queries::delete_web_authn_registration(
                        &mut **transaction,
                        queries::DeleteWebAuthnRegistrationParams {
                            user_id: credential.user_id,
                        },
                    )
                    .await?,
                    "delete webauthn registration",
                )
            })
        })
        .await
    }

    pub async fn create_webauthn_credential(&self, credential: &WebAuthnCredential) -> Result<()> {
        queries::create_web_authn_credential(
            &self.pool,
            queries::CreateWebAuthnCredentialParams {
                id: &credential.id,
                user_id: credential.user_id,
                name: &credential.name,
                credential: &credential.credential,
            },
        )
        .await
        .context("create webauthn credential")
    }

    pub async fn webauthn_credentials_by_user_id(&self, user_id: i64) -> Result<Vec<WebAuthnCredential>> {
        queries::web_authn_credentials_by_user_id(&self.pool, queries::WebAuthnCredentialsByUserIdParams { user_id })
            .await
            .context("load webauthn credentials")
    }

    pub async fn update_webauthn_credential(
        &self,
        id: &str,
        credential_json: &str,
        last_used_at: DateTime<Utc>,
    ) -> Result<()> {
        queries::update_web_authn_credential(
            &self.pool,
            queries::UpdateWebAuthnCredentialParams {
                credential: credential_json,
                last_used_at: Some(last_used_at.into()),
                id,
            },
        )
        .await
        .context("update webauthn credential")
    }

    pub async fn save_login_session_webauthn(&self, id: &str, session: &str) -> Result<()> {
        queries::save_login_session_web_authn(
            &self.pool,
            queries::SaveLoginSessionWebAuthnParams {
                webauthn_session: Some(session),
                id,
            },
        )
        .await
        .context("save webauthn login")
    }

    pub async fn rename_webauthn_credential(&self, id: &str, user_id: i64, name: &str) -> Result<bool> {
        Ok(queries::rename_web_authn_credential(
            &self.pool,
            queries::RenameWebAuthnCredentialParams { name, id, user_id },
        )
        .await?
            == 1)
    }

    pub async fn delete_webauthn_credential(&self, id: &str, user_id: i64) -> Result<bool> {
        Ok(
            queries::delete_web_authn_credential(&self.pool, queries::DeleteWebAuthnCredentialParams { id, user_id })
                .await?
                == 1,
        )
    }

    pub async fn save_webauthn_registration(&self, registration: &WebAuthnRegistration) -> Result<()> {
        queries::save_web_authn_registration(
            &self.pool,
            queries::SaveWebAuthnRegistrationParams {
                user_id: registration.user_id,
                name: &registration.name,
                session: &registration.session,
                expires_at: registration.expires_at,
            },
        )
        .await
        .context("save webauthn registration")
    }

    pub async fn webauthn_registration(&self, user_id: i64) -> Result<Option<WebAuthnRegistration>> {
        queries::web_authn_registration_opt(&self.pool, queries::WebAuthnRegistrationParams { user_id })
            .await
            .context("load webauthn registration")
    }

    pub async fn delete_webauthn_registration(&self, user_id: i64) -> Result<()> {
        affected(
            queries::delete_web_authn_registration(&self.pool, queries::DeleteWebAuthnRegistrationParams { user_id })
                .await?,
            "delete webauthn registration",
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::database::dbtest;

    #[tokio::test]
    async fn round_trips_generated_webauthn_models() {
        let pool = dbtest::open().await.unwrap();
        let user = queries::insert_user(
            &pool,
            queries::InsertUserParams {
                email: "person@example.com",
                role: "admin",
                invited_by: None,
            },
        )
        .await
        .unwrap();
        let store = Store::new(pool);
        let now = "2026-09-06T12:00:00Z".parse::<DateTime<Utc>>().unwrap();
        let registration = WebAuthnRegistration {
            user_id: user.id,
            name: "Passkey".to_owned(),
            session: "{\"challenge\":\"test\"}".to_owned(),
            expires_at: now.into(),
        };
        store.save_webauthn_registration(&registration).await.unwrap();
        assert_eq!(
            store.webauthn_registration(user.id).await.unwrap().unwrap(),
            registration
        );

        let credential = WebAuthnCredential {
            id: "credential".to_owned(),
            user_id: user.id,
            name: registration.name,
            credential: "{\"ID\":\"test\"}".to_owned(),
            created_at: now.into(),
            last_used_at: None,
        };
        store.consume_webauthn_registration(&credential).await.unwrap();
        assert!(store.webauthn_registration(user.id).await.unwrap().is_none());
        let loaded = store.webauthn_credentials_by_user_id(user.id).await.unwrap();
        assert_eq!(loaded.len(), 1);
        assert_eq!(loaded[0].id, credential.id);
        assert_eq!(loaded[0].name, credential.name);
        assert_eq!(loaded[0].user_id, credential.user_id);
        assert_eq!(loaded[0].credential, credential.credential);
        assert_eq!(loaded[0].last_used_at, None);

        store
            .update_webauthn_credential(&credential.id, "{\"ID\":\"updated\"}", now)
            .await
            .unwrap();
        let updated = store.webauthn_credentials_by_user_id(user.id).await.unwrap();
        assert_eq!(updated[0].created_at, loaded[0].created_at);
        assert_eq!(updated[0].credential, "{\"ID\":\"updated\"}");
        assert_eq!(updated[0].last_used_at, Some(now.into()));
    }
}
